src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -325,6 +358,522 @@ CreateInsertSelectRouterPlan(Query *originalQuery,
+													 eqClass->equivalenceId);
+
+				/*
+				 * It seems inefficient to start from the beginning.
+				 * But, we should somehow restart from the beginning to test that
+				 * whether the already skipped ones are equal or not.
+				 */
+				equivalenceClassIndex = 0;
+
+				break;
+			}
+		}
+	}
+
+	return mainEquivalenceClass;
+}
+
+
+/*
+ * Per each relation restriction. get the eq_classes that is in the plannerInfo.
+ * Then, add all the equivalent members of a class to a var equivalence class.
+ */
+static List *
+GenerateVarEquivalencesForRelationRestrictions(
+	RelationRestrictionContext *restrictionContext)
+{
+	List *varEquivalences = NIL;
+	ListCell *relationRestrictionCell = NULL;
+
+	foreach(relationRestrictionCell, restrictionContext->relationRestrictionList)
+	{
+		RelationRestriction *relationRestriction = lfirst(relationRestrictionCell);
+		List *equivalenceClasses = relationRestriction->plannerInfo->eq_classes;
+		ListCell *equivilanceClassCell = NULL;
+		PlannerInfo *plannerInfo = relationRestriction->plannerInfo;
+
+		foreach(equivilanceClassCell, equivalenceClasses)
+		{
+			EquivalenceClass *plannerEqClass = lfirst(equivilanceClassCell);
+			ListCell *equivilanceMemberCell = NULL;
+			VarEquivalenceClass *varEquivalance = palloc0(sizeof(VarEquivalenceClass));
+			varEquivalance->equivalenceId = varEquivalenceId++;
+
+			foreach(equivilanceMemberCell, plannerEqClass->ec_members)
+			{
+				EquivalenceMember *equivalenceMember = lfirst(equivilanceMemberCell);
+				Expr *equivalenceExpr = equivalenceMember->em_expr;
+				Var *expressionVar = NULL;
+
+				if (IsA(equivalenceExpr, Param))
+				{
+					List *parentParamList = relationRestriction->parentPlannerParamList;
+					Param *equivalenceParam = (Param *) equivalenceExpr;
+
+					expressionVar = GetVarFromAssignedParam(parentParamList,
+															equivalenceParam);
+					if (expressionVar)
+					{
+						AddToVarEquivalenceClass(relationRestriction->parentPlannerInfo,
+												 expressionVar,
+												 &varEquivalance);
+					}
+				}
+				else if (IsA(equivalenceExpr, Var))
+				{
+					expressionVar = (Var *) equivalenceExpr;
+					AddToVarEquivalenceClass(plannerInfo, expressionVar, &varEquivalance);
+				}
+			}
+
+			varEquivalences = lappend(varEquivalences, varEquivalance);
+		}
+	}
+
+	return varEquivalences;
+}
+
+
+/*
+ * GetVarFromAssignedParam returns the Var that is assigned to the given
+ * plannerParam if its kind is PARAM_EXEC.
+ *
+ * If the paramkind is not equal to PARAM_EXEC the function returns NULL. Similarly,
+ * if there is no var that the given param is assigned to, the function returns NULL.
+ */
+static Var *
+GetVarFromAssignedParam(List *parentPlannerParamList, Param *plannerParam)
+{
+	Var *assignedVar = NULL;
+	ListCell *plannerParameterCell = NULL;
+
+	Assert(plannerParam != NULL);
+
+	/* we're only interested in parameters that Postgres added for execution */
+	if (plannerParam->paramkind != PARAM_EXEC)
```

Comment:
Add regression tests with parametrized prepared statements. (w/wout `PARAM_EXEC` in the query)